### DISCUSSION TITLE
offJacobian of coupled aux variable v in PorousFlowHeatMassTransfer.C

### DISCUSSION THREAD

**Discussion by Bearunner** [idaholab/moose#26810] [Q&A Modules: General] [answered]
Dear all,

I find that the offJacobian of aux variable **'_v(coupledValue("v")),'** needs to be defined in [PorousFlowHeatMassTransfer.C](https://github.com/idaholab/moose/blob/next/modules/porous_flow/src/kernels/PorousFlowHeatMassTransfer.C) see bold in below.

Real
PorousFlowHeatMassTransfer::computeQpResidual()
{
  return _coef_var[_qp] * (_u[_qp] - _v[_qp]) * _test[_i][_qp];
}

Real
PorousFlowHeatMassTransfer::jac(unsigned int jvar) const
{
  if (jvar == _var.number())
    return _coef_var[_qp] * _phi[_j][_qp] * _test[_i][_qp];
  **else if (jvar == _v_var)
    return -_coef_var[_qp] * _phi[_j][_qp] * _test[_i][_qp];**
  return 0.0;
}

So now if I set v as an aux variable and equals to a constant value in the input file. Do I still need the OffJacobian calculation with respect to v as shown above? And why?

Thanks for your help.

B

**Comment by Bearunner** [comment]
hello, @GiudGiud , could you please help me with this? Thanks

**Reply by GiudGiud** [reply]
What are v and u here?

The kernel is for the equation that is associated with the degrees of the variable parameter

If v is coupled to that variable but it is not a nonlinear variable, then there is no off diagonal

**Reply by Bearunner** [reply]
Hi Guillaume, thanks for your tips.

> 1.What are v and u here?

u is the nonlinear and main variable of this kernel. v is the coupled aux variable.

> 2. If v is coupled to that variable but it is not a nonlinear variable, then there is no off diagonal

If the coupled aux variable v is from another app by the [MultiAppShapeEvaluationTransfer] or [MultiAppNearestNodeTransfer]. The v is updated at (every picard interation) and time-step. 

Do we need the off-diagonal of coupled aux variable v in the kernel of nonlinear and main variable u?

Thanks again,

B

**Reply by GiudGiud** [reply, marked as answer]
If v comes from another application, it is not being solved for in the nonlinear system by Petsc. So there's no row/column corresponding to DoFs for v and we do not need to add entries in the Jacobian for that coupling
